File: skills/xcrawl-crawl/scripts/crawl_create_and_poll.py
#!/usr/bin/env python3
import argparse
import json
import os
import sys
import time

import requests


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Create and poll an Xcrawl crawl task")
    parser.add_argument("url", nargs="?", default=None, help="Entry URL (positional shortcut)")
    parser.add_argument("--url", dest="url_opt", help="Entry URL")
    parser.add_argument("--limit", type=int, default=100, help="Crawler limit for shortcut mode")
    parser.add_argument("--max-depth", type=int, default=2, help="Crawler max_depth for shortcut mode")
    parser.add_argument(
        "--formats",
        default="markdown",
        help="Comma-separated output formats for shortcut mode",
    )
    parser.add_argument("--payload-file", help="Path to raw JSON payload file")
    parser.add_argument("--payload-json", help="Raw JSON payload string")
    parser.add_argument("--interval", type=float, default=2.0, help="Polling interval in seconds")
    parser.add_argument("--max-attempts", type=int, default=30, help="Polling max attempts")
    return parser.parse_args()


def load_payload(args: argparse.Namespace) -> dict:
    if args.payload_file and args.payload_json:
        raise SystemExit("Use only one of --payload-file or --payload-json")

    if args.payload_file:
        with open(args.payload_file, "r", encoding="utf-8") as f:
            return json.load(f)

    if args.payload_json:
        return json.loads(args.payload_json)

    url = args.url_opt or args.url or "https://example.com"
    formats = [item.strip() for item in args.formats.split(",") if item.strip()]
    return {
        "url": url,
        "crawler": {
            "limit": args.limit,
            "max_depth": args.max_depth,
        },
        "output": {
            "formats": formats or ["markdown"],
        },
    }


def main() -> None:
    api_key = os.getenv("XCRAWL_API_KEY")
    if not api_key:
        raise SystemExit("XCRAWL_API_KEY is required")

    args = parse_args()
    payload = load_payload(args)
    headers = {
        "Authorization": f"Bearer {api_key}",
        "Content-Type": "application/json",
    }

    create_resp = requests.post(
        "https://run.xcrawl.com/v1/crawl",
        headers=headers,
        json=payload,
        timeout=60,
    )
    create_text = create_resp.text
    print(create_text)
    if not create_resp.ok:
        sys.exit(1)

    try:
        create_data = create_resp.json()
    except ValueError as exc:
        raise SystemExit(f"create response is not JSON: {exc}")

    crawl_id = create_data.get("crawl_id")
    if not crawl_id:
        raise SystemExit("crawl_id missing in create response")

    for attempt in range(args.max_attempts):
        result_resp = requests.get(
            f"https://run.xcrawl.com/v1/crawl/{crawl_id}",
            headers={"Authorization": f"Bearer {api_key}"},
            timeout=60,
        )
        result_text = result_resp.text
        if not result_resp.ok:
            print(result_text)
            sys.exit(1)

        try:
            result_data = result_resp.json()
        except ValueError as exc:
            raise SystemExit(f"result response is not JSON: {exc}")

        status = result_data.get("status")
        if status in {"completed", "failed"}:
            print(result_text)
            return

        if attempt < args.max_attempts - 1:
            time.sleep(args.interval)

    raise SystemExit(
        f"polling timed out after {args.max_attempts} attempts; crawl_id={crawl_id}"
    )


if __name__ == "__main__":
    main()

File: skills/xcrawl-crawl/scripts/crawl_create_and_poll.js
#!/usr/bin/env node
const fs = require("node:fs");

function parseArgs(argv) {
  const args = { _: [] };
  for (let i = 0; i < argv.length; i += 1) {
    const token = argv[i];
    if (!token.startsWith("--")) {
      args._.push(token);
      continue;
    }
    const eqIndex = token.indexOf("=");
    if (eqIndex !== -1) {
      const key = token.slice(2, eqIndex);
      args[key] = token.slice(eqIndex + 1);
      continue;
    }
    const key = token.slice(2);
    const next = argv[i + 1];
    if (next && !next.startsWith("--")) {
      args[key] = next;
      i += 1;
    } else {
      args[key] = "true";
    }
  }
  return args;
}

function buildPayload(args) {
  if (args["payload-file"] && args["payload-json"]) {
    throw new Error("Use only one of --payload-file or --payload-json");
  }
  if (args["payload-file"]) {
    return JSON.parse(fs.readFileSync(args["payload-file"], "utf8"));
  }
  if (args["payload-json"]) {
    return JSON.parse(args["payload-json"]);
  }

  const formats = String(args.formats || "markdown")
    .split(",")
    .map((item) => item.trim())
    .filter(Boolean);

  return {
    url: args.url || args._[0] || "https://example.com",
    crawler: {
      limit: Number(args.limit || 100),
      max_depth: Number(args["max-depth"] || 2),
    },
    output: {
      formats: formats.length ? formats : ["markdown"],
    },
  };
}

async function main() {
  const apiKey = process.env.XCRAWL_API_KEY;
  if (!apiKey) {
    console.error("XCRAWL_API_KEY is required");
    process.exit(1);
  }

  const args = parseArgs(process.argv.slice(2));
  const payload = buildPayload(args);
  const interval = Number(args.interval || 2);
  const maxAttempts = Number(args["max-attempts"] || 30);

  const createResp = await fetch("https://run.xcrawl.com/v1/crawl", {
    method: "POST",
    headers: {
      Authorization: `Bearer ${apiKey}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  const createText = await createResp.text();
  process.stdout.write(`${createText}\n`);
  if (!createResp.ok) {
    process.exit(1);
  }

  let createData;
  try {
    createData = JSON.parse(createText);
  } catch (err) {
    throw new Error(`create response is not JSON: ${err.message}`);
  }

  const crawlId = createData.crawl_id;
  if (!crawlId) {
    throw new Error("crawl_id missing in create response");
  }

  for (let i = 0; i < maxAttempts; i += 1) {
    const resultResp = await fetch(`https://run.xcrawl.com/v1/crawl/${crawlId}`, {
      headers: { Authorization: `Bearer ${apiKey}` },
    });

    const resultText = await resultResp.text();
    if (!resultResp.ok) {
      process.stdout.write(`${resultText}\n`);
      process.exit(1);
    }

    let resultData;
    try {
      resultData = JSON.parse(resultText);
    } catch (err) {
      throw new Error(`result response is not JSON: ${err.message}`);
    }

    const status = resultData.status;
    if (status === "completed" || status === "failed") {
      process.stdout.write(`${resultText}\n`);
      return;
    }

    if (i < maxAttempts - 1) {
      await new Promise((resolve) => setTimeout(resolve, interval * 1000));
    }
  }

  throw new Error(`polling timed out after ${maxAttempts} attempts; crawl_id=${crawlId}`);
}

main().catch((err) => {
  console.error(err.message || err);
  process.exit(1);
});
